cardano_node_tests/tests/test_plutus_v2_spend_build.py

```diff
@@ -1462,3 +1498,464 @@ def test_v1_attached_v2_reference(
+        reference_input_utxo = cluster.get_utxo(txin=reference_txin)
+
+        assert not reference_input_utxo, f"The reference input was spent `{reference_txin}`"
+
+        # TODO check command 'transaction view' bug on cardano-node 4045
+
+
+@pytest.mark.testnets
+class TestNegativeReferenceInputs:
+    """Tests for Tx with reference inputs that are expected to fail."""
+
+    @allure.link(helpers.get_vcs_link())
+    def test_invalid_reference_input(
+        self, cluster: clusterlib.ClusterLib, payment_addrs: List[clusterlib.AddressRecord]
+    ):
+        """Test use an invalid reference input when unlock some funds.
+
+        Expect failure
+        """
+        __: Any  # mypy workaround
+        temp_template = common.get_test_id(cluster)
+
+        plutus_op = PLUTUS_OP_ALWAYS_SUCCEEDS
+
+        # for mypy
+        assert plutus_op.execution_cost
+        assert plutus_op.datum_file
+        assert plutus_op.redeemer_cbor_file
+
+        # create the necessary Tx outputs
+
+        script_utxos, collateral_utxos, __, __ = _build_fund_script(
+            temp_template=temp_template,
+            cluster=cluster,
+            payment_addr=payment_addrs[0],
+            dst_addr=payment_addrs[1],
+            plutus_op=plutus_op,
+            use_inline_datum=False,
+        )
+
+        #  spend the "locked" UTxO
+
+        plutus_txins = [
+            clusterlib.ScriptTxIn(
+                txins=script_utxos,
+                script_file=plutus_op.script_file,
+                collaterals=collateral_utxos,
+                execution_units=(
+                    plutus_op.execution_cost.per_time,
+                    plutus_op.execution_cost.per_space,
+                ),
+                redeemer_cbor_file=plutus_op.redeemer_cbor_file,
+                datum_file=plutus_op.datum_file,
+            )
+        ]
+
+        tx_files_redeem = clusterlib.TxFiles(
+            signing_key_files=[payment_addrs[1].skey_file],
+        )
+        txouts_redeem = [
+            clusterlib.TxOut(address=payment_addrs[1].address, amount=-1),
+        ]
+
+        # create an invalid reference input
+        reference_input = clusterlib.UTXOData(
+            utxo_hash=collateral_utxos[0].utxo_hash,
+            utxo_ix=200,
+            amount=2_000_000,
+            address=payment_addrs[0].address,
+        )
+
+        with pytest.raises(clusterlib.CLIError) as excinfo:
+            cluster.build_tx(
+                src_address=payment_addrs[0].address,
+                tx_name=f"{temp_template}_step2",
+                tx_files=tx_files_redeem,
+                readonly_reference_txins=[reference_input],
+                txouts=txouts_redeem,
+                script_txins=plutus_txins,
+            )
+        err_str = str(excinfo.value)
+        assert "TranslationLogicMissingInput" in err_str, err_str
+
+    @allure.link(helpers.get_vcs_link())
+    def test_reference_spent_output(
```

Comment:
How does this test differ from `test_invalid_reference_input`? Both seem to work with non-existent UTxO as txin.

Reply:
Yeah when I was writing that made sense, the first one with a utxo that never exists the second one with a utxo that existed but was already spent. But you are totally right, in the end is the same :) I will drop the first one.